File: src/agent/UnityConnection/UnityManager.ts
/**
 * Improved Unity Manager for WebSocket connection management.
 *
 * Integrates all the best practices:
 * - Session management with monotonic takeover
 * - Heartbeat/keepalive with compilation-aware suspension
 * - Message routing with ACK support
 * - Command/response correlation
 * - Graceful reconnection handling
 */

import { WebSocket } from 'ws';
import { randomUUID } from 'crypto';
import { createLogger } from './config';
import {
    ExtendedConnection,
    MovesiaMessage,
    ConnectionState,
    ConnectionSource,
    UnityManagerConfig,
    DEFAULT_UNITY_MANAGER_CONFIG,
    CloseCode,
    SessionEntry,
    createExtendedConnection
} from './types';
import { UnitySessionManager } from './sessions';
import { HeartbeatManager } from './heartbeat';
import { MessageRouter, CommandRouter, RouterCallbacks } from './router';
import { sendToClient, sendWelcome } from './transport';

const logger = createLogger('movesia.unity');

// =============================================================================
// Types
// =============================================================================

/** Callback for connection state changes. */
export type ConnectionChangeCallback = (connected: boolean) => Promise<void>;

/** Callback for domain events. */
export type DomainEventCallback = (msg: MovesiaMessage) => Promise<void>;

/** Interrupt manager interface (for async operations). */
export interface InterruptManager {
    resumeAll(): Promise<void>;
}

// =============================================================================
// Unity Manager
// =============================================================================

/**
 * Manages WebSocket connections from Unity Editor.
 *
 * Features:
 * - Single active connection per project/session
 * - Automatic takeover of older connections
 * - Heartbeat with compilation-aware suspension
 * - Command/response correlation for tool calls
 * - Interrupt support for async operations
 *
 * Usage:
 *     const manager = new UnityManager();
 *
 *     // In WebSocket endpoint
 *     await manager.handleConnection(websocket);
 *
 *     // From tools
 *     const result = await manager.sendAndWait("query_hierarchy", { path: "/" });
 */
export class UnityManager {
    readonly config: UnityManagerConfig;
    private _interruptManager?: InterruptManager;
    private _onDomainEvent?: DomainEventCallback;

    // Session management
    private _sessions: UnitySessionManager;

    // Heartbeat management
    private _heartbeat: HeartbeatManager;

    // Message routing
    private _router: MessageRouter;

    // Command routing for request/response
    private _commandRouter: CommandRouter;

    // Current connection tracking (for single Unity connection)
    private _currentWs?: WebSocket;
    private _currentConnection?: ExtendedConnection;
    private _currentSession?: string;

    // Connection change callbacks
    private _connectionCallbacks: ConnectionChangeCallback[] = [];

    // Pending commands awaiting responses (keyed by message ID)
    private _pendingCommands: Map<string, {
        resolve: (value: Record<string, unknown>) => void;
        reject: (reason: Error) => void;
    }> = new Map();

    constructor(options: {
        interruptManager?: InterruptManager;
        config?: Partial<UnityManagerConfig>;
        onDomainEvent?: DomainEventCallback;
    } = {}) {
        this.config = { ...DEFAULT_UNITY_MANAGER_CONFIG, ...options.config };
        this._interruptManager = options.interruptManager;
        this._onDomainEvent = options.onDomainEvent;

        // Session management
        this._sessions = new UnitySessionManager();

        // Heartbeat management
        this._heartbeat = new HeartbeatManager({
            config: this.config.heartbeat,
            getConnections: () => this._getAllConnections(),
            sendPing: (ws, cid) => this._sendPing(ws, cid),
            closeConnection: (ws, code, reason) => this._closeConnection(ws, code, reason)
        });

        // Message routing
        const routerCallbacks: RouterCallbacks = {
            suspendHeartbeat: (ms) => this._heartbeat.suspend(ms),
            onDomainEvent: (msg) => this._handleDomainEvent(msg),
            sendToClient: (ws, msg) => this._sendToWebsocket(ws, msg),
            onCompilationStarted: (cid) => this._onCompilationStarted(cid),
            onCompilationFinished: (cid) => this._onCompilationFinished(cid)
        };
        this._router = new MessageRouter(routerCallbacks);

        // Command routing
        this._commandRouter = new CommandRouter();
    }

    // =========================================================================
    // Public API
    // =========================================================================

    /**
     * Handle a new Unity WebSocket connection.
     *
     * This is the main entry point called from the WebSocket endpoint.
     *
     * @param websocket - WebSocket connection
     * @param sessionId - Session identifier (from query param or handshake)
     * @param connSeq - Connection sequence number for takeover logic
     */
    async handleConnection(
        websocket: WebSocket,
        providedSessionId?: string,
        connSeq: number = 0
    ): Promise<void> {
        // Generate connection ID
        const cid = this._generateCid();

        // Session and connSeq come from URL query params (no handshake needed)
        const sessionId: string = providedSessionId ?? randomUUID();

        // Create connection metadata
        const connection = createExtendedConnection(cid, {
            session: sessionId,
            connSeq
        });

        // Try to accept the session
        const decision = await this._sessions.accept(
            sessionId,
            connSeq,
            connection,
            websocket
        );

        if (!decision.accept) {
            logger.info(`Rejecting connection [${cid}]: ${decision.reason}`);
            websocket.close(CloseCode.DUPLICATE_SESSION, decision.reason ?? 'duplicate session');
            return;
        }

        // Supersede old connection if needed
        if (decision.supersede) {
            try {
                decision.supersede.close(CloseCode.SUPERSEDED, 'superseded by newer connection');
            } catch (error) {
                logger.debug(`Error closing superseded connection: ${error}`);
            }
        }

        // Update current connection
        this._currentWs = websocket;
        this._currentConnection = connection;
        this._currentSession = sessionId;
        connection.state = ConnectionState.OPEN;

        // Start heartbeat if not running
        this._heartbeat.start();

        // Notify connection change
        await this._notifyConnectionChange(true);

        // Send welcome message
        await sendWelcome(websocket, {
            cid,
            session: sessionId,
            server_version: '2.0.0'
        });

        const shortSession = sessionId.substring(0, 8);
        logger.info(`Unity connected [${cid}] session=${shortSession}`);

        // Set up event handlers
        websocket.on('message', async (data) => {
            try {
                await this._handleMessage(websocket, connection, data);
            } catch (error) {
                logger.error(`Error handling message [${cid}]`, error as Error);
            }
        });

        websocket.on('close', async (code, _reason) => {
            logger.info(`Unity disconnected [${cid}] code=${code}`);
            await this._cleanupConnection(websocket, connection, sessionId);
        });

        websocket.on('error', async (error) => {
            logger.error(`Unity connection error [${cid}]`, error);
            await this._cleanupConnection(websocket, connection, sessionId);
        });
    }

    /**
     * Send a command to Unity and wait for response.
     *
     * @param commandType - Type of command (e.g., "query_hierarchy")
     * @param params - Command parameters
     * @param timeout - Timeout in seconds (defaults to config)
     * @returns Response body from Unity
     * @throws Error if no Unity connection
     * @throws Error if response times out
     */
    async sendAndWait(
        commandType: string,
        params: Record<string, unknown> = {},
        timeout?: number
    ): Promise<Record<string, unknown>> {
        if (!this._currentWs || !this._currentConnection) {
            throw new Error('No Unity connection available');
        }

        const timeoutMs = (timeout ?? this.config.commandTimeout) * 1000;

        // Create and send command
        const msg = MovesiaMessage.create(
            commandType,
            params,
            ConnectionSource.VSCODE,
            this._currentSession
        );

        // Register for response using message ID (Unity echoes this back)
        const responsePromise = new Promise<Record<string, unknown>>((resolve, reject) => {
            this._pendingCommands.set(msg.id, { resolve, reject });
        });

        // Set up timeout
        const timeoutId = setTimeout(() => {
            const pending = this._pendingCommands.get(msg.id);
            if (pending) {
                this._pendingCommands.delete(msg.id);
                pending.reject(new Error(`Command ${commandType} timed out after ${timeoutMs}ms`));
            }
        }, timeoutMs);

        logger.info(`Registered pending command: msg.id=${msg.id}`);

        try {
            await sendToClient(this._currentWs, msg.toDict());
            logger.info(`Sent command ${commandType} [msg.id=${msg.id}]`);

            return await responsePromise;
        } finally {
            clearTimeout(timeoutId);
            this._pendingCommands.delete(msg.id);
        }
    }

    /**
     * Check if Unity is currently connected.
     */
    get isConnected(): boolean {
        return (
            this._currentWs !== undefined &&
            this._currentConnection !== undefined &&
            this._currentConnection.state === ConnectionState.OPEN
        );
    }

    /**
     * Get current Unity project path.
     */
    get currentProject(): string | undefined {
        return this._currentConnection?.projectPath;
    }

    /**
     * Check if Unity is currently compiling.
     */
    get isCompiling(): boolean {
        return this._currentConnection?.isCompiling ?? false;
    }

    /**
     * Get number of active connections.
     */
    get connectionCount(): number {
        return this._sessions.size;
    }

    /**
     * Register callback for connection state changes.
     */
    onConnectionChange(callback: ConnectionChangeCallback): void {
        this._connectionCallbacks.push(callback);
    }

    /**
     * Close all Unity connections.
     */
    async closeAll(): Promise<void> {
        this._heartbeat.stop();
        this._commandRouter.cancelAll();

        const sessions = await this._sessions.getAllSessions();
        for (const [_sessionId, entry] of sessions) {
            try {
                entry.websocket.close(CloseCode.GOING_AWAY, 'server shutdown');
            } catch {
                // Ignore close errors
            }
        }

        await this._sessions.clearAll();
        this._currentWs = undefined;
        this._currentConnection = undefined;
        this._currentSession = undefined;
    }

    // =========================================================================
    // Private Implementation
    // =========================================================================

    /**
     * Handle incoming message.
     */
    private async _handleMessage(
        websocket: WebSocket,
        connection: ExtendedConnection,
        data: Buffer | ArrayBuffer | Buffer[]
    ): Promise<void> {
        const rawData = Buffer.isBuffer(data) ? data : Buffer.from(data as ArrayBuffer);

        // Route through message router
        const msg = await this._router.handleMessage(websocket, connection, rawData);

        if (!msg) {
            return;
        }

        // Check if this is a response to a pending command (matched by message ID)
        // Unity echoes back the original message ID in its response
        const pending = this._pendingCommands.get(msg.id);
        if (pending) {
            this._pendingCommands.delete(msg.id);
            pending.resolve(msg.body);
        }
    }

    /**
     * Clean up after connection closes.
     */
    private async _cleanupConnection(
        websocket: WebSocket,
        connection: ExtendedConnection,
        sessionId: string
    ): Promise<void> {
        connection.state = ConnectionState.CLOSED;

        // Clear from sessions
        await this._sessions.clearIfMatch(sessionId, websocket);

        // Clear current connection if it matches
        if (this._currentWs === websocket) {
            this._currentWs = undefined;
            this._currentConnection = undefined;
            this._currentSession = undefined;
        }

        // Cancel pending commands
        for (const [_requestId, pending] of this._pendingCommands) {
            pending.reject(new Error('Connection closed'));
        }
        this._pendingCommands.clear();

        // Notify connection change
        await this._notifyConnectionChange(false);

        // Stop heartbeat if no more connections
        if (this._sessions.size === 0) {
            this._heartbeat.stop();
        }

        logger.info(`Cleaned up connection [${connection.cid}]`);
    }

    /**
     * Forward domain events to subscribers.
     */
    private async _handleDomainEvent(msg: MovesiaMessage): Promise<void> {
        if (this._onDomainEvent) {
            try {
                await this._onDomainEvent(msg);
            } catch (error) {
                logger.error('Error in domain event handler', error as Error);
            }
        }
    }

    /**
     * Handle Unity compilation start.
     */
    private async _onCompilationStarted(cid: string): Promise<void> {
        logger.info(`Unity compilation started [${cid}]`);

        // Cancel pending commands (they'll fail anyway)
        for (const [_requestId, pending] of this._pendingCommands) {
            pending.reject(new Error('Compilation started'));
        }
        this._pendingCommands.clear();
    }

    /**
     * Handle Unity compilation finish.
     */
    private async _onCompilationFinished(cid: string): Promise<void> {
        logger.info(`Unity compilation finished [${cid}]`);

        // Resume any interrupted operations
        if (this._interruptManager) {
            try {
                await this._interruptManager.resumeAll();
            } catch (error) {
                logger.error('Error resuming interrupts', error as Error);
            }
        }
    }

    /**
     * Notify all connection change callbacks.
     */
    private async _notifyConnectionChange(connected: boolean): Promise<void> {
        for (const callback of this._connectionCallbacks) {
            try {
                await callback(connected);
            } catch (error) {
                logger.error('Error in connection change callback', error as Error);
            }
        }
    }

    /**
     * Get all connections for heartbeat manager.
     */
    private async _getAllConnections(): Promise<Map<string, SessionEntry>> {
        return this._sessions.getAllSessions();
    }

    /**
     * Send ping to a connection.
     */
    private async _sendPing(ws: WebSocket, cid: string): Promise<void> {
        const msg = MovesiaMessage.create('hb', {}, ConnectionSource.VSCODE);
        try {
            await sendToClient(ws, msg.toDict());
        } catch (error) {
            logger.debug(`Failed to send ping to [${cid}]: ${error}`);
        }
    }

    /**
     * Close a WebSocket connection.
     */
    private async _closeConnection(
        ws: WebSocket,
        code: number,
        reason: string
    ): Promise<void> {
        try {
            ws.close(code, reason);
        } catch (error) {
            logger.debug(`Error closing connection: ${error}`);
        }
    }

    /**
     * Send message to a WebSocket.
     */
    private async _sendToWebsocket(
        ws: WebSocket,
        message: Record<string, unknown>
    ): Promise<void> {
        await sendToClient(ws, message);
    }

    /**
     * Generate a short connection ID.
     */
    private _generateCid(): string {
        const chars = 'abcdefghijklmnopqrstuvwxyz0123456789';
        let result = '';
        for (let i = 0; i < 8; i++) {
            result += chars.charAt(Math.floor(Math.random() * chars.length));
        }
        return result;
    }
}

// =============================================================================
// Factory Function
// =============================================================================

/**
 * Create a Unity manager with default configuration.
 */
export function createUnityManager(options: {
    interruptManager?: InterruptManager;
    onDomainEvent?: DomainEventCallback;
} = {}): UnityManager {
    return new UnityManager(options);
}

File: agent-python-(deprecated)/unity/unity_manager.py
"""
Improved Unity Manager for WebSocket connection management.

Integrates all the best practices:
- Session management with monotonic takeover
- Heartbeat/keepalive with compilation-aware suspension
- Message routing with ACK support
- Command/response correlation
- Graceful reconnection handling
"""

import asyncio
import logging
import uuid
from typing import Optional, Callable, Awaitable, Any, Dict
from contextlib import asynccontextmanager

from fastapi import WebSocket, WebSocketDisconnect

from .types import (
    ExtendedConnection,
    MovesiaMessage,
    ConnectionState,
    ConnectionSource,
    UnityManagerConfig,
    HeartbeatConfig,
    CloseCode
)
from .sessions import UnitySessionManager
from .heartbeat import HeartbeatManager
from .router import MessageRouter, CommandRouter, RouterCallbacks
from .transport import send_to_client, send_welcome, send_error, send_command


logger = logging.getLogger("movesia.unity")


class UnityManager:
    """
    Manages WebSocket connections from Unity Editor.
    
    Features:
    - Single active connection per project/session
    - Automatic takeover of older connections
    - Heartbeat with compilation-aware suspension
    - Command/response correlation for tool calls
    - Interrupt support for async operations
    
    Usage:
        manager = UnityManager(interrupt_manager)
        
        # In WebSocket endpoint
        await manager.handle_connection(websocket)
        
        # From tools
        result = await manager.send_and_wait("query_hierarchy", path="/")
    """
    
    def __init__(
        self,
        interrupt_manager: Any = None,
        config: Optional[UnityManagerConfig] = None,
        on_domain_event: Optional[Callable[[MovesiaMessage], Awaitable[None]]] = None
    ):
        """
        Initialize Unity manager.
        
        Args:
            interrupt_manager: Manager for async interrupts (for domain reload)
            config: Configuration options
            on_domain_event: Callback for domain events from Unity
        """
        self.config = config or UnityManagerConfig()
        self._interrupt_manager = interrupt_manager
        self._on_domain_event = on_domain_event
        
        # Session management
        self._sessions = UnitySessionManager()
        
        # Heartbeat management
        self._heartbeat = HeartbeatManager(
            config=self.config.heartbeat,
            get_connections=self._get_all_connections,
            send_ping=self._send_ping,
            close_connection=self._close_connection
        )
        
        # Message routing
        router_callbacks = RouterCallbacks(
            suspend_heartbeat=self._heartbeat.suspend,
            on_domain_event=self._handle_domain_event,
            send_to_client=self._send_to_websocket,
            on_compilation_started=self._on_compilation_started,
            on_compilation_finished=self._on_compilation_finished
        )
        self._router = MessageRouter(callbacks=router_callbacks)
        
        # Command routing for request/response
        self._command_router = CommandRouter()
        
        # Current connection tracking (for single Unity connection)
        self._current_ws: Optional[WebSocket] = None
        self._current_connection: Optional[ExtendedConnection] = None
        self._current_session: Optional[str] = None
        
        # Connection change callbacks
        self._connection_callbacks: list[Callable[[bool], Awaitable[None]]] = []

        # Pending commands awaiting responses (keyed by message ID)
        self._pending_commands: Dict[str, asyncio.Future] = {}

    # =========================================================================
    # Public API
    # =========================================================================
    
    async def handle_connection(
        self,
        websocket: WebSocket,
        session_id: Optional[str] = None,
        conn_seq: int = 0
    ) -> None:
        """
        Handle a new Unity WebSocket connection.
        
        This is the main entry point called from the WebSocket endpoint.
        
        Args:
            websocket: FastAPI WebSocket
            session_id: Session identifier (from query param or handshake)
            conn_seq: Connection sequence number for takeover logic
        """
        # Accept the WebSocket connection
        await websocket.accept()

        # Generate connection ID
        cid = self._generate_cid()

        # Session and conn_seq come from URL query params (no handshake needed)
        # Project path and unity version can be sent later via "hello" message if needed
        if not session_id:
            session_id = str(uuid.uuid4())
        
        # Create connection metadata
        connection = ExtendedConnection(
            cid=cid,
            session=session_id,
            conn_seq=conn_seq,
        )

        # Try to accept the session
        decision = await self._sessions.accept(
            session_id=session_id,
            conn_seq=conn_seq,
            connection=connection,
            websocket=websocket,
        )
        
        if not decision.accept:
            logger.info(f"Rejecting connection [{cid}]: {decision.reason}")
            await websocket.close(
                code=CloseCode.DUPLICATE_SESSION,
                reason=decision.reason or "duplicate session"
            )
            return
        
        # Supersede old connection if needed
        if decision.supersede:
            try:
                old_ws = decision.supersede
                await old_ws.close(
                    code=CloseCode.SUPERSEDED,
                    reason="superseded by newer connection"
                )
            except Exception as e:
                logger.debug(f"Error closing superseded connection: {e}")
        
        # Update current connection
        self._current_ws = websocket
        self._current_connection = connection
        self._current_session = session_id
        connection.state = ConnectionState.OPEN
        
        # Start heartbeat if not running
        self._heartbeat.start()
        
        # Notify connection change
        await self._notify_connection_change(True)
        
        # Send welcome message
        await send_welcome(websocket, {
            "cid": cid,
            "session": session_id,
            "server_version": "2.0.0"
        })
        
        short_session = session_id[:8] if session_id else "none"
        logger.info(f"Unity connected [{cid}] session={short_session}")
        
        # Main message loop
        try:
            await self._message_loop(websocket, connection)
        except WebSocketDisconnect as e:
            logger.info(f"Unity disconnected [{cid}]")
        except Exception as e:
            logger.error(f"Unity connection error [{cid}]: {e}", exc_info=True)
        finally:
            await self._cleanup_connection(websocket, connection, session_id)
    
    async def send_and_wait(
        self,
        command_type: str,
        timeout: Optional[float] = None,
        **kwargs
    ) -> dict:
        """
        Send a command to Unity and wait for response.
        
        Args:
            command_type: Type of command (e.g., "query_hierarchy")
            timeout: Timeout in seconds (defaults to config)
            **kwargs: Command parameters
            
        Returns:
            Response body from Unity
            
        Raises:
            RuntimeError: If no Unity connection
            asyncio.TimeoutError: If response times out
        """
        if not self._current_ws or not self._current_connection:
            raise RuntimeError("No Unity connection available")
        
        timeout = timeout or self.config.command_timeout

        # Create and send command
        msg = MovesiaMessage.create(
            msg_type=command_type,
            body=kwargs,  # No need for request_id - we use msg.id for correlation
            source=ConnectionSource.VSCODE,
            session=self._current_session
        )

        # Register for response using message ID (Unity echoes this back)
        future: asyncio.Future = asyncio.get_event_loop().create_future()
        self._pending_commands[msg.id] = future
        logger.info(f"📤 Registered pending command: msg.id={msg.id}, pending_commands={list(self._pending_commands.keys())}")

        try:
            await send_to_client(self._current_ws, msg.to_dict())

            logger.info(f"Sent command {command_type} [msg.id={msg.id}]")

            return await asyncio.wait_for(future, timeout)

        except asyncio.TimeoutError:
            logger.warning(f"Command {command_type} timed out after {timeout}s")
            raise
        finally:
            self._pending_commands.pop(msg.id, None)
    
    @property
    def is_connected(self) -> bool:
        """Check if Unity is currently connected."""
        return (
            self._current_ws is not None and 
            self._current_connection is not None and
            self._current_connection.state == ConnectionState.OPEN
        )
    
    @property
    def current_project(self) -> Optional[str]:
        """Get current Unity project path."""
        if self._current_connection:
            return self._current_connection.project_path
        return None
    
    @property
    def is_compiling(self) -> bool:
        """Check if Unity is currently compiling."""
        if self._current_connection:
            return self._current_connection.is_compiling
        return False
    
    @property
    def connection_count(self) -> int:
        """Get number of active connections."""
        return self._sessions.size
    
    def on_connection_change(
        self, 
        callback: Callable[[bool], Awaitable[None]]
    ) -> None:
        """Register callback for connection state changes."""
        self._connection_callbacks.append(callback)
    
    async def close_all(self) -> None:
        """Close all Unity connections."""
        self._heartbeat.stop()
        await self._command_router.cancel_all()
        
        sessions = await self._sessions.get_all_sessions()
        for session_id, entry in sessions.items():
            try:
                await entry.websocket.close(
                    code=CloseCode.GOING_AWAY,
                    reason="server shutdown"
                )
            except Exception:
                pass
        
        await self._sessions.clear_all()
        self._current_ws = None
        self._current_connection = None
        self._current_session = None
    
    # =========================================================================
    # Private Implementation
    # =========================================================================

    async def _message_loop(
        self,
        websocket: WebSocket,
        connection: ExtendedConnection
    ) -> None:
        """Main loop for receiving and processing messages."""
        while True:
            # Receive message
            data = await websocket.receive_text()


            # Route through message router
            msg = await self._router.handle_message(websocket, connection, data)

            if msg is None:
                continue

            # Check if this is a response to a pending command (matched by message ID)
            # Unity echoes back the original message ID in its response
            if msg.id in self._pending_commands:
                future = self._pending_commands.get(msg.id)
                if future and not future.done():
                    future.set_result(msg.body)
    
    async def _cleanup_connection(
        self,
        websocket: WebSocket,
        connection: ExtendedConnection,
        session_id: str
    ) -> None:
        """Clean up after connection closes."""
        connection.state = ConnectionState.CLOSED
        
        # Clear from sessions
        await self._sessions.clear_if_match(session_id, websocket)
        
        # Clear current connection if it matches
        if self._current_ws is websocket:
            self._current_ws = None
            self._current_connection = None
            self._current_session = None
        
        # Cancel pending commands
        for request_id, future in list(self._pending_commands.items()):
            if not future.done():
                future.set_exception(RuntimeError("Connection closed"))
        self._pending_commands.clear()
        
        # Notify connection change
        await self._notify_connection_change(False)
        
        # Stop heartbeat if no more connections
        if self._sessions.size == 0:
            self._heartbeat.stop()
        
        logger.info(f"Cleaned up connection [{connection.cid}]")
    
    async def _handle_domain_event(self, msg: MovesiaMessage) -> None:
        """Forward domain events to subscribers."""
        if self._on_domain_event:
            try:
                await self._on_domain_event(msg)
            except Exception as e:
                logger.error(f"Error in domain event handler: {e}")
    
    async def _on_compilation_started(self, cid: str) -> None:
        """Handle Unity compilation start."""
        logger.info(f"Unity compilation started [{cid}]")
        
        # Cancel pending commands (they'll fail anyway)
        for request_id, future in list(self._pending_commands.items()):
            if not future.done():
                future.set_exception(RuntimeError("Compilation started"))
        self._pending_commands.clear()
    
    async def _on_compilation_finished(self, cid: str) -> None:
        """Handle Unity compilation finish."""
        logger.info(f"Unity compilation finished [{cid}]")
        
        # Resume any interrupted operations
        if self._interrupt_manager:
            try:
                await self._interrupt_manager.resume_all()
            except Exception as e:
                logger.error(f"Error resuming interrupts: {e}")
    
    async def _notify_connection_change(self, connected: bool) -> None:
        """Notify all connection change callbacks."""
        for callback in self._connection_callbacks:
            try:
                await callback(connected)
            except Exception as e:
                logger.error(f"Error in connection change callback: {e}")
    
    async def _get_all_connections(self) -> Dict[str, Any]:
        """Get all connections for heartbeat manager."""
        return await self._sessions.get_all_sessions()
    
    async def _send_ping(self, ws: WebSocket, cid: str) -> None:
        """Send ping to a connection."""
        msg = MovesiaMessage.create(
            msg_type="hb",
            body={},
            source=ConnectionSource.VSCODE
        )
        try:
            await send_to_client(ws, msg.to_dict())
        except Exception as e:
            logger.debug(f"Failed to send ping to [{cid}]: {e}")
    
    async def _close_connection(
        self, 
        ws: WebSocket, 
        code: int, 
        reason: str
    ) -> None:
        """Close a WebSocket connection."""
        try:
            await ws.close(code=code, reason=reason)
        except Exception as e:
            logger.debug(f"Error closing connection: {e}")
    
    async def _send_to_websocket(self, ws: WebSocket, message: dict) -> None:
        """Send message to a WebSocket."""
        await send_to_client(ws, message)
    
    @staticmethod
    def _generate_cid() -> str:
        """Generate a short connection ID."""
        import random
        import string
        return ''.join(random.choices(string.ascii_lowercase + string.digits, k=8))


# Convenience function for backwards compatibility with existing code
def create_unity_manager(
    interrupt_manager: Any = None,
    on_domain_event: Optional[Callable[[MovesiaMessage], Awaitable[None]]] = None
) -> UnityManager:
    """Create a Unity manager with default configuration."""
    return UnityManager(
        interrupt_manager=interrupt_manager,
        on_domain_event=on_domain_event
    )
